packages/coreutils/src/signal.ts

```diff
@@ -0,0 +1,35 @@
+import { PromiseDelegate } from '@lumino/coreutils';
+import { ISignal } from '@lumino/signaling';
+
+/**
+ * Convert a signal into a promise for the first emitted value.
+ *
+ * @param signal - The signal we are listening to.
+ * @param timeout - Timeout to wait for signal in ms (not timeout if not defined or 0)
+ *
+ * @returns a Promise that resolves with a `(sender, args)` pair.
+ */
+export function signalToPromise<T, U>(
+  signal: ISignal<T, U>,
+  timeout?: number
+): Promise<[T, U]> {
+  const waitForSignal = new PromiseDelegate<[T, U]>();
+
+  function slot(sender: T, args: U) {
+    waitForSignal.resolve([sender, args]);
+
+    cleanup();
+  }
+  signal.connect(slot);
+
+  if ((timeout ?? 0) > 0) {
+    setTimeout(() => {
+      waitForSignal.reject(`Signal not emitted within ${timeout} ms.`);
```

Comment:
Do we need to disconnect the `slot` after timeout?

Reply:
Indeed we should.